Source/Core/DiscIO/NANDContentLoader.cpp

```diff
@@ -248,92 +201,64 @@ bool CNANDContentLoader::Initialize(const std::string& name)
+    tmd_file.ReadBytes(bytes.data(), bytes.size());
+    m_tmd.SetBytes(std::move(bytes));
 
-  std::copy(&tmd[0], &tmd[TMD_HEADER_SIZE], m_TMDHeader);
-  std::copy(&tmd[0x180], &tmd[0x180 + TMD_VIEW_SIZE], m_TMDView);
-
-  m_TitleVersion = Common::swap16(&tmd[0x01DC]);
-  m_NumEntries = Common::swap16(&tmd[0x01DE]);
-  m_BootIndex = Common::swap16(&tmd[0x01E0]);
-  m_TitleID = Common::swap64(&tmd[0x018C]);
-  m_IosVersion = Common::swap16(&tmd[0x018A]);
-  m_Country = static_cast<u8>(m_TitleID & 0xFF);
-
-  if (m_Country == 2)  // SYSMENU
-    m_Country = GetSysMenuRegion(m_TitleVersion);
+    m_ticket = FindSignedTicket(m_tmd.GetTitleId());
+  }
 
-  InitializeContentEntries(tmd, decrypted_title_key, data_app);
+  InitializeContentEntries(data_app);
   return true;
 }
 
-void CNANDContentLoader::InitializeContentEntries(const std::vector<u8>& tmd,
-                                                  const std::vector<u8>& decrypted_title_key,
-                                                  const std::vector<u8>& data_app)
+void CNANDContentLoader::InitializeContentEntries(const std::vector<u8>& data_app)
 {
-  m_Content.resize(m_NumEntries);
+  if (!m_ticket.IsValid())
+  {
+    ERROR_LOG(IOS_ES, "No valid ticket for title %016" PRIx64, m_tmd.GetTitleId());
+    return;
+  }
 
-  std::array<u8, 16> iv;
-  u32 data_app_offset = 0;
+  const std::vector<IOS::ES::Content> contents = m_tmd.GetContents();
+  m_Content.resize(contents.size());
 
+  u32 data_app_offset = 0;
+  const std::vector<u8> title_key = m_ticket.GetTitleKey();
   CSharedContent shared_content{Common::FromWhichRoot::FROM_SESSION_ROOT};
 
-  for (u32 i = 0; i < m_NumEntries; i++)
+  for (size_t i = 0; i < contents.size(); ++i)
   {
-    const u32 entry_offset = 0x24 * i;
-
-    SNANDContent& content = m_Content[i];
-    content.m_ContentID = Common::swap32(&tmd[entry_offset + 0x01E4]);
-    content.m_Index = Common::swap16(&tmd[entry_offset + 0x01E8]);
-    content.m_Type = Common::swap16(&tmd[entry_offset + 0x01EA]);
-    content.m_Size = static_cast<u32>(Common::swap64(&tmd[entry_offset + 0x01EC]));
-
-    const auto header_begin = std::next(tmd.begin(), entry_offset + 0x01E4);
-    const auto header_end = std::next(header_begin, ArraySize(content.m_Header));
-    std::copy(header_begin, header_end, content.m_Header);
-
-    const auto hash_begin = std::next(tmd.begin(), entry_offset + 0x01F4);
-    const auto hash_end = std::next(hash_begin, ArraySize(content.m_SHA1Hash));
-    std::copy(hash_begin, hash_end, content.m_SHA1Hash);
+    const auto& content = contents.at(i);
 
     if (m_IsWAD)
     {
-      u32 rounded_size = Common::AlignUp(content.m_Size, 0x40);
-
-      iv.fill(0);
-      std::copy(&tmd[entry_offset + 0x01E8], &tmd[entry_offset + 0x01E8 + 2], iv.begin());
+      // The content index is used as IV (2 bytes); the remaining 14 bytes are zeroes.
+      std::array<u8, 16> iv{};
+      iv[0] = (content.index >> 8) & 0xFF;
```

Comment:
These both need to be static_casted to `u8`, otherwise you're truncating `int` sized values.